src/ncp/ncp_base_ftd.cpp

```diff
@@ -114,47 +89,28 @@ otError NcpBase::GetPropertyHandler_THREAD_CHILD_TABLE(uint8_t aHeader, spinel_p
                         childInfo.mFullNetworkData
                     );
 
-        SuccessOrExit(
-            error = OutboundFrameFeedPacked(
-                        SPINEL_DATATYPE_STRUCT_S(
-                            SPINEL_DATATYPE_EUI64_S         // EUI64 Address
-                            SPINEL_DATATYPE_UINT16_S        // Rloc16
-                            SPINEL_DATATYPE_UINT32_S        // Timeout
-                            SPINEL_DATATYPE_UINT32_S        // Age
-                            SPINEL_DATATYPE_UINT8_S         // Network Data Version
-                            SPINEL_DATATYPE_UINT8_S         // Link Quality In
-                            SPINEL_DATATYPE_INT8_S          // Average RSS
-                            SPINEL_DATATYPE_UINT8_S         // Mode (flags)
-                            SPINEL_DATATYPE_INT8_S          // Most recent RSS
-                        ),
-                        childInfo.mExtAddress.m8,
-                        childInfo.mRloc16,
-                        childInfo.mTimeout,
-                        childInfo.mAge,
-                        childInfo.mNetworkDataVersion,
-                        childInfo.mLinkQualityIn,
-                        childInfo.mAverageRssi,
-                        modeFlags,
-                        childInfo.mLastRssi
-                    ));
-    }
+        SuccessOrExit(error = mEncoder.OpenStruct());
 
-    SuccessOrExit(error = OutboundFrameSend());
+        SuccessOrExit(error = mEncoder.WriteEui64(childInfo.mExtAddress));
+        SuccessOrExit(error = mEncoder.WriteUint16(childInfo.mRloc16));
+        SuccessOrExit(error = mEncoder.WriteUint32(childInfo.mTimeout));
+        SuccessOrExit(error = mEncoder.WriteUint32(childInfo.mAge));
+        SuccessOrExit(error = mEncoder.WriteUint8(childInfo.mNetworkDataVersion));
+        SuccessOrExit(error = mEncoder.WriteUint8(childInfo.mLinkQualityIn));
+        SuccessOrExit(error = mEncoder.WriteInt8(childInfo.mAverageRssi));
+        SuccessOrExit(error = mEncoder.WriteUint8(modeFlags));
+        SuccessOrExit(error = mEncoder.WriteInt8(childInfo.mLastRssi));
+
+        SuccessOrExit(error = mEncoder.CloseStruct());
+    }
```

Comment:
This scheme does appear to be an improvement over the previous. I like how the format and value are brought together on the same line.  This allows us to avoid counting lines to learn which format goes with which value.  

What is lacking in both of these schemes is a formal contract regarding the order of the elements in the buffer.  There does not exist a template which can be used by both sender and receiver of this packet to guarantee that the formats and order used by both will be the same.  For instance, for one to write the host side code that will unpack this message, one must look at the NCP side code, here and manually construct a block of code that will unpack the message.  One cannot change the order of these contents without ensuring that the order is changed on all the hosts that have implemented a solution to the older order/format.

The only way I can see to solve this, would be to move the encoding and decoding of the message out of ncp_... and into spinel itself or perhaps some new source file that could be shared by both the host and NCP.

Reply:
Fully agree... longer-term plan/goal is to have a stand-alone implementation of a `SpinelPropertyHandler` (which would contain both encoding and decoding logic for properties) and can be re-used across platforms